### DISCUSSION TITLE
Would my filter list be a good feature for uBlock Origin and its community?

### DISCUSSION THREAD

**Discussion by ghost** [uBlockOrigin/uAssets#23649] [Q&A] [answered]
Hello,

I have curated a list of 2000+ scam videos and channels that spread malware on youtube in my github repo. I was wondering if it was possible to add it to one of the filter lists as a custom import list would not block these links to the scam videos and channels. 

The repo is [here](https://github.com/Incrypters/YT-ScamDatabase)

Thank you and have a nice day!

**Comment by garry-ut99** [comment]
> as a custom import list would not block these links to the scam videos and channels.

Works with `$doc` or `$all`, example:

```
||youtube.com/channel/UCK5Es63vf6gAgflsR2Ase5Q^$all
||youtube.com/watch?v=VtSEHB5XEMg^$all
```
Documentation: 
https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#document
https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#all

**Reply by ghost** [reply]
Dynamic navigation with XHR? I would use IDs only, with doc and xhr option.

**Reply by garry-ut99** [reply]
Good idea, I also saw some video requests smuggled as `image`.

<hr>

> Polycarbohydrate : Thanks for the info!

I see you have added `$doc`, but it seems `$all` or `$doc,xhr,image` will work better, because `$doc` only works when:
- entering URL directly into browser address bar 
- or opening a video link in a new tab
- but not when watching some youtube video
and then clicking thumbnail of other video
because they use `xhr` requests

So for example you should put in your list (maybe only one asterisk will be enough):

```
*UCK5Es63vf6gAgflsR2Ase5Q*$doc,xhr,image
*VtSEHB5XEMg*$doc,xhr,image
```

or

```
*UCK5Es63vf6gAgflsR2Ase5Q*$all
*VtSEHB5XEMg*$all
```

**But sometimes videos still does play (by clicking a video thumbnail, even new videos, which means not cached yet)**

**Reply by ghost** [reply]
Did what you suggested, made an asterisk before the ID and changed to *$all. Thanks.

**Reply by ghost** [reply]
Looks like clicking on the video does not block even with *UCK5Es63vf6gAgflsR2Ase5Q*$all

Might be like you said, it hasnt cached yet.

**Reply by Yuki2718** [reply]
`*VtSEHB5XEMg*$all` such filters won't be tokenized and affects performance by the numbers of them.

**Reply by garry-ut99** [reply]
> Incrypters : Looks like clicking on the video does not block even with `UCK5Es63vf6gAgflsR2Ase5Q$all`
> Incrypters : Might be like you said, it hasnt cached yet.

I didn't mean it needs to be cached to be blocked, I only meant that it wasn't served from the cache, as sometimes files which are served from browser cache can bypass filters, so I just excluded such possibility from being the reason.

Maybe the real reason why videos sometimes still get through is that youtube videos are served also via requests which don't contain the unique video ID.

> Yuki2718 : `*VtSEHB5XEMg*$all` such filters won't be tokenized and affects performance by the numbers of them.
> garry-ut99 : (maybe only one asterisk will be enough)

Ok, ( https://github.com/gorhill/uBlock/wiki/Filter-Performance#narrowing-options-for-network-filters ) ,  in such case an asterisk at the beginning is sufficient then.

**Reply by Yuki2718** [reply]
Single asterisk would still prevent tokenization. What are you going to block? Better to add individual rules like you originally said
> ||youtube.com/channel/UCK5Es63vf6gAgflsR2Ase5Q^$all
> ||youtube.com/watch?v=VtSEHB5XEMg^$all

I don't believe yt has infinite numbers of patterns.

**Reply by garry-ut99** [reply]
> What are you going to block?

I tried wildcards because:
- a) videos were sometimes still playing when clicking at video thumbnails on the right (and they still do, even with a wildcard applied, so it seems that idea with adding wildcard didn't work to resolve this issue)
- b) wildcard blocks also video thumbnails, and not only video, which is desired and useful, because if the purpose is to block fake/scam videos, then blocking theirs thumbnails is desired as well


Your filters:
- still "don't block videos when clicking at video thumbnails on the right" either (a)
- and additionaly leak video thumbnails (b)

> Better to add individual rules like you originally said

but I do agree that if a single wildcard still does make a difference in terms of performance, then it's better to split them to more efficient specific wildcard-less counterparts.

**Reply by stephenhawk8054** [reply]
Video/channel ID (whether with wildcard or not) won't help when users click on videos inside youtube page, because when you do that, it generates `?prettyPrint` XHR POST connections and sends the ID via Request body. You might need something like

```adb
youtube.com##+js(trusted-replace-fetch-response, /.+?videoId":"(?:nwoCoj4j14g|VtSEHB5XEMg)".+/, , /player?)
```

but it can only prevent the video from loading and cannot block the page, maybe combining it with something like

```adb
youtube.com##ytd-watch-flexy:is([video-id="nwoCoj4j14g"],[video-id="VtSEHB5XEMg"]):remove()
```

will help

**Reply by garry-ut99** [reply]
The scriptlet works fine.

The cosmetic filter without `:remove()` works fine as well.

But `:remove()` part causes the cosmetic filter malfunctions sometimes:
- sometimes doesn't work
- sometimes does work, but after pressing "back" arrow in browser, the previous video loads fine (can hear the sound), but the cosmetic filter still hides the whole page

All the problems are gone after removing `:remove()` part, but curious why you added `:remove()`, and whether any problems will appear without it.

**Reply by stephenhawk8054** [reply]
Just wanted to remove the node, but if it causes issue, yeah you can remove it.

Anyways, what I mean is currently there's no ways to do it via network filters alone and you'll need more complicated approach for this issue.

**Reply by garry-ut99** [reply]
Ok, I'm just trying to help OP, hence the final example, for a single video:
```
||youtube.com/channel/UCK5Es63vf6gAgflsR2Ase5Q$all,domain=youtube.com
||youtube.com/watch?v=VtSEHB5XEMg$all,domain=youtube.com
||i.ytimg.com/vi/VtSEHB5XEMg$image,domain=youtube.com
youtube.com##ytd-compact-video-renderer:has([href="/watch?v=VtSEHB5XEMg"])
youtube.com##ytd-compact-radio-renderer:has([href^="/watch?v=VtSEHB5XEMg"])
youtube.com##+js(trusted-replace-fetch-response, /.+?videoId":"(?:VtSEHB5XEMg|SomeOtherVideoID)".+/, , /player?)
youtube.com##ytd-watch-flexy:is([video-id="VtSEHB5XEMg"],[video-id="SomeOtherVideoID"])
```

It still might be not complete because maybe need more network/cosmetic filters needed to hide all thumbnails (for ex. on main search page).

Also I'm not sure about this one: `youtube.com##ytd-compact-radio-renderer:has([href^="/watch?v=VtSEHB5XEMg"])` as this thumbnail leads to a video but at the same time to a video playlist compilation.

Personally I use BlockTube extension to block unwanted videos and channels:
- https://chromewebstore.google.com/detail/blocktube/bbeaicapbccfllodepmimpkgecanonai
- https://addons.mozilla.org/en-CA/firefox/addon/blocktube/

**Comment by ghost** [comment, marked as answer]
Commit history is short. Promote, maintain, build user base.

**Reply by garry-ut99** [reply]
I agree.

**Comment by Yuki2718** [comment]
And why `^` at the end of all rules? Read documentation.
